### ISSUE TITLE
Project compiler constants not applied to packages

### ISSUE THREAD

**Issue by fafalone** [twinbasic/twinbasic#1890]
**Describe the bug**
In project settings, there's an option to specify compiler constants. Prior to Beta 424, these were applied to packages referenced by the project, however now they're not applied.

**To Reproduce**
For example, in a project with WinDevLib added, set compiler constant `WINDEVLIB_LITE=1`, then in local project code, `Dim x As Long = L2_REASON_CODE_GROUP_SIZE`. It should give an error, as that constant is within a module, wdAPIWLAN.twin, that should have been excluded `#If WINDEVLIB_LITE <> 0`.

**Expected behavior**
Compiler constants should be applied to packages a project has added.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows 10 
 - twinBASIC compiler version 0.15.592

**Additional context**

**Comment by fafalone** [comment]
I'd think a warning about the consequences for packages in the compiler constants definition area would be sufficient... something like "WARNING: For packages, any compiler constants defined which affect the package will apply to all components making use of it."

At least in the interim pending possible architectural changes to support different in the future, presumably sometime after v1.0.

**Comment by WaynePhillipsEA** [comment]
Lets say you've got a project that references Package 1 and Package 2, and both of those packages each reference Package 3 individually.  Package 1 has different compilation args defined than Package 2.   Which compilation args will apply when Package 3 gets loaded (once only)?   

I can only really see it working in its current form if we do indeed limit compilation arguments to only be allowed at the project level, and not the package level.  That may well be a suitable/reasonable compromise for now.

**Comment by fafalone** [comment]
Definitely an improvement over not applying them at all; I think package makers will understand the need for the limitation at present; and it shouldn't break any of the existing packages now.

**Comment by bclothier** [comment]
One caveat.... I think the package should be able to provide a default value for the constants. Meaning, if a package has `HOT_DOG_STAND_MODE_ENABLED = 1` in its project, and it's then referenced in another project, and that top level project does _not_ have `HOT_DOG_STAND_MODE_ENABLED` defined, it should default to `1`, not `0`.

**Comment by fafalone** [comment]
That sounds problematic.

Say someone makes a package using WDL and uses `WINDEVLIB_LITE = 1`, wiping out 90% of API defs. 

The people who add it at a higher level are going to be responsible for figuring that out then overriding it back to 0?

**Comment by fafalone** [comment]
The burden here should fall on package developers, not top level users. 

You're also risking creating compatibility problems, where people can't use some package because it breaks or seriously compromises another package. For now, package makers should just have to accept they can't configure dependent packages that way.

**Comment by WaynePhillipsEA** [comment]
> Say someone makes a package using WDL and uses WINDEVLIB_LITE = 1, wiping out 90% of API defs.

No, that's not what I was envisaging.  The only "inheritance" we can offer is from the root project compilation args.  In other words; a package would be unable to affect any referenced packages that it uses - only the root project compilation args can be "passed on" down the chain.  You will of course be able to set compilation arguments within your own package, but 1) they won't affect any referenced sub-packages, and 2) they can be overridden by the root project compilation args.

**Comment by fafalone** [comment]
Oh ok, sorry I misunderstood.

**Comment by WaynePhillipsEA** [comment]
> Consider WINDEVLIB_NO_QUADLI; if one package uses it then it would force a higher level user to either not use that package or accept a huge number of broken UDTs, since enabling that option would create invalid syntax in the package choosing to use it if overridden.

I can't see us being able to offer any form of compilation args "inheritance" other than what I described in my post above, due to the limitations of packages only being compiled once.  It does mean that if you apply the "wrong" compilation args at the project level you risk breaking an indirect package in the chain.   For now, I guess it would be the package owners responsibility to list any useful and compatible compilation args that can be set at the project level.  e.g. if a package uses WinDevLib and works fine with WINDEVLIB_LITE, then you could list that as being a compatible argument (which will work provided all other referenced packages that use WinDevLib are also similarly compatible).

**Comment by fafalone** [comment]
Ok as I understood your last post, there is no inheritance save for the top level project. i.e. the only way someone could activate a WDL flag would be to either put it in the top level project, or build a modified WinDevLib.twinpack with it in the settings from the .twinproj.

Project A
--Package B
----WinDevLib

Package B's .twinproj wouldn't affect WDL for Project A, is that correct?

If so then yes; it should be up to package makers to understand while they could list compatible flags for packages their own ues, they couldn't expect them to be set or not.

**Comment by bclothier** [comment]
FWIW, if we do go down that path, the package makers would probably have to avoid using generic names like `DEBUG` or other common flag names to avoid unexpected behavior when the top level projects uses those common names. 

I don't know if it's feasible to have a project disambiguated name (e.g. `WinDevLib.DEBUG` vs. `PackageB.DEBUG` vs. `DEBUG` which applies at top-level project only) but if that is feasible, that would be one way to have more control over how the flag should "inherit" without having to insist that every package maker use longer names like `WINDEVLIB_DEBUG` to avoid being conflated with the `DEBUG` at the top level.

**Comment by fafalone** [comment]
That should be part of responsible package development, yes. Knowing it wasn't a top level project is why mine are prefixed `WINDEVLIB_`. You need to expect that not all compiler constants will be for you.

(In fact in WDL there's even one constant you have to specifically enable for the package to put a bunch of rarely used short generic names into use in code, and others were renamed because the SDK itself used short, generic names for less rarely used stuff where some headers conflict with others). 

Qualifying the names... maybe we don't need such long unique prefixes as I went with, but I do think package makers should just operate under the assumption their constants need to be fairly unique vs shifting that burden to users of the package; packages should be as easy to use as possible. 

At some point one of us could update the documentation on package creation to explain all this.
